rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -110,64 +138,175 @@ public void close(ISourceLocation loc) {
+            //     notifications) monotonically increases with each `calculate`
+            //     call.
+
+            var delayed = CompletableFuture.delayedExecutor(delay.toMillis(), TimeUnit.MILLISECONDS, exec);
+            CompletableFuture.runAsync(() -> {
+                // If no new call to `calculate` has been made after `delay` has
+                // passed (i.e., `lastVersion` hasn't changed in the meantime),
+                // then run the calculation. Else, abandon this calculation.
+                if (latestVersion.get() == version) {
+                    calculation.run();
+                }
+            }, delayed);
         }
 
-        private void reportTypeCheckerMessages(List<Diagnostic> messages) {
-            typeCheckerMessages = messages;
-            sendDiagnostics();
+        public void calculateAnalyzer(CompletableFuture<Versioned<ITree>> tree, int version, Duration delay) {
+            debounce(version, latestVersionCalculateAnalyzer, delay, () ->
+                analyzer.calculateSummary(tree).thenAcceptIfUninterrupted(messages ->
+                    reportDiagnostics(analyzerDiagnostics, version, messages.get())));
         }
 
-        public void invalidate(boolean isClosing) {
-            summary.invalidate(isClosing);
+        /**
+         * The main complication when running the builder is that it might
+         * produce a subset of the same diagnostics as the analyzer. Thus, as
+         * the latest parser/analyzer/builder diagnostics are always reported
+         * together, a diff needs to be computed of the analyzer diagnostics and
+         * the builder diagnostics to avoid reporting duplicates produced by
+         * both analyzer and builder.
+         */
+        public void calculateBuilder(CompletableFuture<Versioned<ITree>> tree) {
+
+            // Schedule the analyzer. This is *always* needed, because the
+            // latest result of `calculateAnalyzer` may be for a syntax tree
+            // with a greater version than parameter `tree` (because
+            // `calculateAnalyzer` has debouncing), or it may be interrupted due
+            // to later change (which should not affect the builder).
+            builderAnalysis = analyzer.calculateMessages(tree);
+
+            // Schedule the builder and use exactly the same syntax tree as the
+            // analyzer. In this way, a reliable diff of the analyzer
+            // diagnostics and the builder diagnostics can be computed (by
+            // removing the former from the latter).
+            builderBuild = builder.calculateSummary(tree);
+
+            // Only if neither the analyzer nor the builder was interrupted,
+            // report diagnostics. Otherwise, *no* diagnostics are reported
+            // (instead of reporting an empty list of diagnostics).
+            builderAnalysis.thenAcceptBothIfUninterrupted(builderBuild,
+
+                // When the analyzer or the builder is interrupted, its future
```

Comment:
is this comment still relevant, doesn't `acceptBothIfUninterrupted` already take care of that?